refactor: Shared to use internal mutability

akaladarshi opened the pull request:
Fixes: #170 

This PR changes:
- Now the `Shared` struct internally handles the mutex.
- `Shared` struct now has `inner: Arc<Mutex<SharedInner>>`

elenaf9 commented on yamux/src/connection/stream.rs:
```diff
@@ -316,37 +317,47 @@ impl AsyncRead for Stream {
-            if chunk.is_empty() {
-                shared.buffer.pop();
-                continue;
+        let can_read = self.shared.with_mut(|inner| {
```
Can't we just return here directly the `Poll` result?
That way we could move the whole bottom part of the function as it was inside this closure.

akaladarshi replied:
It's now using simple `lock` instead of `with_mut`, so no need for this anymore.

elenaf9 commented on yamux/src/connection/stream.rs:
```diff
@@ -393,8 +415,9 @@ impl AsyncWrite for Stream {
         if frame.header().flags().contains(ACK) {
-            self.shared()
-                .update_state(self.conn, self.id, State::Open { acknowledged: true });
+            self.shared.with_mut(|inner| {
+                inner.update_state(self.conn, self.id, State::Open { acknowledged: true });
+            });
```
```suggestion
            self.shared.update_state(self.conn, self.id, State::Open { acknowledged: true });
```

Isn't there a `update_state` convenience function on `Shared`?

akaladarshi replied:
Done.

elenaf9 commented on yamux/src/connection/stream.rs:
```diff
@@ -429,30 +452,90 @@ impl AsyncWrite for Stream {
+        self.with_mut(|inner| inner.update_state(cid, sid, next))
+    }
+
+    pub fn with_mut<F, R>(&self, f: F) -> R
```
`parking_lot::Mutex` doesn't support re-entrant locking, thus a nested call to this would cause a deadlock. 
To avoid this, the function should take `&mut self`.

akaladarshi replied:
Thanks for pointing this out!

I need a suggestion regarding the design choices in the code snippet below. We are frequently using `Pin<&mut Self> ` in `connection.rs`. So if  we change `fn with
_mut` to take `&mut self` then, we would run into borrow checkers issue.

So to fix it, is the below snippet a good choice or we could do something like take a `mutexGuard` on sharedInner or is there any better?.

Here’s the code for reference (stream.rs -> fn poll_next):

```rust
        let conn_id = self.conn.clone();
        let stream_id = self.id.clone();
        let shared = &mut self.shared;
        let poll_state = shared.with_mut(|inner| {
            if let Some(bytes) = inner.buffer.pop() {
                let off = bytes.offset();
                let mut vec = bytes.into_vec();
                if off != 0 {
                    // This should generally not happen when the stream is used only as
                    // a `futures::stream::Stream` since the whole point of this impl is
                    // to consume chunks atomically. It may perhaps happen when mixing
                    // this impl and the `AsyncRead` one.
                    log::debug!(
                        "{}/{}: chunk has been partially consumed",
                        conn_id,
                        stream_id,
                    );
                    vec = vec.split_off(off)
                }
                return Poll::Ready(Some(Ok(Packet(vec))));
            }

            // Buffer is empty, let's check if we can expect to read more data.
            if !inner.state.can_read() {
                log::debug!("{}/{}: eof", conn_id, stream_id);
                return Poll::Ready(None); // stream has been reset
            }

            // Since we have no more data at this point, we want to be woken up
            // by the connection when more becomes available for us.
            inner.reader = Some(cx.waker().clone());
            Poll::Pending
        });
```

akaladarshi replied:
Updated the `with_mut` to use `&mut`.

I added another `lock()` function, which will allows us to use simple lock wherever it's necessary and also keeping the `with_mut` so we can use whichever we need.

elenaf9 replied:
> I need a suggestion regarding the design choices in the code snippet below. We are frequently using `Pin<&mut Self> `in `connection.rs.` So if we change `fn with_mut` to take `&mut self` then, we would run into borrow checkers issue.

You could use `Pin::get_mut` and then destructure `Self` to access the distinct fields, which would satisfy the borrow checker:
```rs
        let Self {
            conn, shared, id, ..
        } = self.get_mut();
        shared.with_mut(|inner| {
            inner.update_state(*conn, *id, State::SendClosed);
        });
```

> I added another `lock()` function, which will allows us to use simple lock wherever it's necessary and also keeping the `with_mut` so we can use whichever we need.

`lock` also needs to take `&mut self` for the same reasons.
Generally, I am not a fan of having both `lock` and `with_mut` since they are just two ways of doing the same thing. I am okay with either of them, but we should decide for one.

akaladarshi replied:
Ohh okay so essential same as storing seperately as my example suggest but using `get_mut`.

Then I think we will go with the `with_mut` as it is more dynamic implementation.

elenaf9 commented on yamux/src/connection/stream.rs:
```diff
@@ -429,30 +452,90 @@ impl AsyncWrite for Stream {
             .map_err(|_| self.write_zero_err())?;
-        self.shared()
-            .update_state(self.conn, self.id, State::SendClosed);
+        self.shared.with_mut(|inner| {
+            inner.update_state(self.conn, self.id, State::SendClosed);
+        });
```
Same here: i think we can use directly `Shared::update_state`?

akaladarshi replied:
Done.

elenaf9 reviewed (commented):
Thank you for your PR @akaladarshi! 

Some comments.

akaladarshi commented on yamux/src/connection.rs:
```diff
@@ -628,35 +627,40 @@ impl<T: AsyncRead + AsyncWrite + Unpin> Active<T> {
+                    inner.consume_receive_window(frame.body_len());
+                    inner.buffer.push(frame.into_body());
+                })
             }
```
Reverted the changes to use `self.shared.lock()`.

akaladarshi commented on yamux/src/connection.rs:
```diff
@@ -628,35 +627,40 @@ impl<T: AsyncRead + AsyncWrite + Unpin> Active<T> {
+                    if let Some(w) = inner.reader.take() {
+                        w.wake()
+                    }
+                    Action::None
```
Done.

akaladarshi commented on yamux/src/connection/stream.rs:
```diff
@@ -262,9 +265,7 @@ impl futures::stream::Stream for Stream {
-        let mut shared = self.shared();
-
-        if let Some(bytes) = shared.buffer.pop() {
+        if let Some(bytes) = self.shared.pop_buffer() {
```
Removed it, now using single lock.

akaladarshi commented:
@elenaf9 Thanks for the review 🙏. 

 I have removed the `lock` and directly using `with_mut`, just one small change I have refactored Stream `write_zero_err` fn to be a static as it was just for creating error and it was causing few issues with borrow checkers in `poll_write`.

elenaf9 commented on yamux/src/connection/stream.rs:
```diff
@@ -360,54 +362,65 @@ impl AsyncWrite for Stream {
+
+        log::trace!("{}/{}: write {} bytes", stream.conn, stream.id, n);
 
-        // technically, the frame hasn't been sent yet on the wire but from the perspective of this data structure, we've queued the frame for sending
-        // We are tracking this information:
-        // a) to be consistent with outbound streams
-        // b) to correctly test our behaviour around timing of when ACKs are sent. See `ack_timing.rs` test.
```
I don't think this comment should be removed?

akaladarshi replied:
Done.

elenaf9 reviewed (approved):
One comment, rest LGTM. Thank you for the follow-ups @akaladarshi!

cc @jxs: think this change is an improvement because it hides the mutex-locking logic behind the `Shared` type and prevents deadlocks that would otherwise been possible when trying to locking the Mutex re-entrantly. Are you okay with merging?
